index/mem/index.go

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"crypto/md5"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+const initialSize = 1024 * 1024
+
+// TODO(prateek): investigate impact of native heap
+type memIndex struct {
+	opts   Options
+	hashFn doc.HashFn
+
+	docIDGen *atomic.Uint32
+
+	// hash(doc.ID)  -> internal docID
+	idsLock sync.RWMutex
+	idsMap  map[doc.Hash]index.DocID
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[index.DocID]document
+
+	// fieldName -> (Term -> postingID)
+	fieldLock sync.RWMutex
+	fieldMap  map[string]*termsDict
+
+	// postingID -> bitmap of docIDs
+	postingsList postingsList
+}
+
+type document struct {
+	doc.Document
+	docID      index.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(opts Options) Index {
+	hashFn := func(i doc.ID) doc.Hash {
+		return md5.Sum([]byte(i))
```

Comment:
Perhaps an over optimization, but maybe its worth using a different non-cryptographic hash function? I did [some benchmarks](https://github.com/jeromefroe/golang_benchmarks#non-cryptographic-hash-functions) a while back and md5 is one of the slowest non-cryptographic hash while not offering any benefits in desirable non-cryptographic hash function properties (even distribution, avalanche effect, etc.)

Reply:
Sure, i'll put down a TODO to evaluate perf impact of different hash functions

Reply:
done